hail/src/main/scala/is/hail/expr/ir/ComparisonOp.scala

```diff
@@ -14,28 +14,21 @@ object ComparisonOp {
     if (lt != rt)
       throw new RuntimeException(s"Cannot compare types $lt and $rt")
 
-  val fromStringAndTypes: PartialFunction[(String, Type, Type), ComparisonOp[_]] = {
-    case ("==" | "EQ", t1, t2) =>
-      checkCompatible(t1, t2)
-      EQ(t1, t2)
-    case ("!=" | "NEQ", t1, t2) =>
-      checkCompatible(t1, t2)
-      NEQ(t1, t2)
-    case (">=" | "GTEQ", t1, t2) =>
-      checkCompatible(t1, t2)
-      GTEQ(t1, t2)
-    case ("<=" | "LTEQ", t1, t2) =>
-      checkCompatible(t1, t2)
-      LTEQ(t1, t2)
-    case (">" | "GT", t1, t2) =>
-      checkCompatible(t1, t2)
-      GT(t1, t2)
-    case ("<" | "LT", t1, t2) =>
-      checkCompatible(t1, t2)
-      LT(t1, t2)
-    case ("Compare", t1, t2) =>
-      checkCompatible(t1, t2)
-      Compare(t1, t2)
+  val fromString: PartialFunction[String, ComparisonOp[_]] = {
+    case "==" | "EQ" =>
+      EQ(null, null)
+    case "!=" | "NEQ" =>
+      NEQ(null, null)
+    case ">=" | "GTEQ" =>
+      GTEQ(null, null)
+    case "<=" | "LTEQ" =>
+      LTEQ(null, null)
+    case ">" | "GT" =>
+      GT(null, null)
+    case "<" | "LT" =>
+      LT(null, null)
+    case "Compare" =>
+      Compare(null, null)
   }
```

Comment:
These `ComparisonOp`s have always seemed a little funny to me. I'm not sure what value they add over lifting the operand types into `ApplyComparisonOp` and just using the `CodeOrdering.Op`, ie:
```scala
case class ApplyComparisonOp(op: CodeOrdering.Op, aTy: Type, bTy: Type, a: IR, b: IT)
```
I also wonder what's the point of `ApplyComparisonOp` over `ApplyBinaryOp` (or indeed, `Apply`).

Reply:
Fair questions, but I think outside the scope of this PR. For the first question, I think it's mostly about ease of pattern matching; if an optimization wants to apply only for `<`, you don't want to muck around at the `CodeOrdering` level.

Similar for the second question, it's just an issue of what information is encoded in the IR, and how easy it is for optimizations to extract that information.